### ISSUE TITLE
Regression in Node 14

### ISSUE THREAD

**Issue by curbengh** [hexojs/hexo#4260]
Noticed this issue on my blog, Netlify doesn't upload any file after upgrading Node 14 ([build log](https://app.netlify.com/sites/curben/deploys/5e9fa695dfb3bb0007e16aa1)), and led to an [empty site](https://5e9fa695dfb3bb0007e16aa1--curben.netlify.app/).

A working workaround is to downgrade; I downgraded to [Node 12](https://app.netlify.com/sites/curben/deploys/5ea1746c1472010006c804d5), my blog never had any issue with Node 13, so it looks like a Node 14 issue.

I can replicate this issue in https://github.com/hexojs/site/pull/1395, site does become [empty](https://5ea1799ab2cf4b0007f29443--hexo-site.netlify.app/).

---

Replicated the issue in my workstation, with a dummy blog ([default plugins](https://github.com/hexojs/hexo-starter/blob/259ecc3982e0445075f6c0daaccc24e4f1c7e663/package.json#L15-L23) only). `/public` folder has all the files, but they are **all empty**.

- Install `node_modules` with Node 14, run Hexo with Node 14: :x:
- Install `node_modules` with Node 13, run Hexo with Node 14: :x:
- Install `node_modules` with Node 13, run Hexo with Node 13: :white_check_mark:
- Install `node_modules` with Node 14, run Hexo with Node 13: :white_check_mark:

Node 13 is v[`13.13.0`](https://github.com/nodejs/node/releases/tag/v13.13.0), Node 14 is v[`14.0.0`](https://github.com/nodejs/node/releases/tag/v14.0.0)

@hexojs/core 

---

Tested packages:

:x: [**hexo**](https://travis-ci.com/github/hexojs/hexo/jobs/322788444)
:x: **[hexo-fs](https://travis-ci.com/github/hexojs/hexo-fs/jobs/322524508)**
:x: **[hexo-cli](https://travis-ci.com/github/hexojs/hexo-cli/jobs/322526877)**

:white_check_mark: [warehouse](https://github.com/hexojs/warehouse/pull/79)
:white_check_mark: [hexo-renderer-nunjucks](https://github.com/hexojs/hexo-renderer-nunjucks/pull/2)
:white_check_mark: [hexo-renderer-marked](https://github.com/hexojs/hexo-renderer-marked/pull/146)
:white_check_mark: [hexo-util](https://github.com/hexojs/hexo-util/pull/193)
:white_check_mark: [hexo-generator-feed](https://github.com/hexojs/hexo-generator-feed/pull/135).

---

Edit (30 April 2020): fix released in [hexo-util@2.0.0](https://github.com/hexojs/hexo-util/releases/tag/2.0.0) and [hexo-fs@3.0.0](https://github.com/hexojs/hexo-fs/releases/tag/3.0.0).

There are two options to temporarily fix compatibility issues with Node 14:

1. Downgrade to Node 10 or 12. If you use `.nvmrc` as part of your CI workflow, you need to change the content to `10` or `12`.
2. If you prefer to use Node 14, force upgrade hexo-util and hexo-fs:

``` diff package.json
{
  ...
  "dependencies": {
    "hexo": "^4.0.0",
    "hexo-generator-archive": "^1.0.0",
    "hexo-generator-category": "^1.0.0",
    "hexo-generator-index": "^1.0.0",
    "hexo-generator-tag": "^1.0.0",
    "hexo-renderer-ejs": "^1.0.0",
    "hexo-renderer-stylus": "^1.1.0",
    "hexo-renderer-marked": "^2.0.0",
    "hexo-server": "^1.0.0",
+    "hexo-util": "^2.1.0",
+    "hexo-fs": "^3.0.1"
  }
}
```

```
$ rm -rf node_modules/
$ npm install
```

**Comment by YoshinoriN** [comment]
OK. I found this document [transform.destroy([error])](https://nodejs.org/api/stream.html#stream_transform_destroy_error)

>  Implementors should not override this method, but instead implement readable._destroy(). The default implementation of _destroy() for Transform also emit 'close' unless emitClose is set in false.

But, [hexo-util@ChachStream](https://github.com/hexojs/hexo-util/blob/5b14fd2df882cc3d740f6204441824546b660600/lib/cache_stream.js#L21-L23) seems override `destory` method. Is this correct?

**Comment by curbengh** [comment]
> Node.js v14 has many changes about [stream](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V14.md). I think this issue caused by these changes.

I looked around the https://github.com/nodejs/node PRs, the stream changes already included on Node 13, but all hexo-related packages have no issue with it.

**Comment by jiangtj** [comment]
https://github.com/hexojs/hexo/issues/4257 This problem is also caused by v14

**Comment by SukkaW** [comment]
It seems that all problem is related with `hexo-fs`.

**Comment by stevenjoezhang** [comment]
Yes, it seems that most of the error messages come from this commit

nodejs/node@a13500f#diff-0034beae2e92ed0b59570df638dda305

**Comment by SukkaW** [comment]
The fix has already implemented in https://github.com/hexojs/hexo-fs/pull/60

**Comment by stevenjoezhang** [comment]
Tests show that `cacheStream.destroys` in Node.js 14 will be triggered twice, which causes all content to be empty.
Renaming these two `destroy` methods can solve the problem.

https://github.com/hexojs/hexo-util/blob/31f74a5296d74339b76909a05ee38d8bb38560bf/lib/cache_stream.js#L21
https://github.com/hexojs/hexo/blob/b60f815b975178ee1c907a92e5a1403830e56f37/lib/plugins/console/generate.js#L68

**Comment by curbengh** [comment]
> OK. I found this document [transform.destroy([error])](https://nodejs.org/api/stream.html#stream_transform_destroy_error)
> 
> > Implementors should not override this method, but instead implement readable._destroy(). The default implementation of _destroy() for Transform also emit 'close' unless emitClose is set in false.
> 
> But, [hexo-util@ChachStream](https://github.com/hexojs/hexo-util/blob/5b14fd2df882cc3d740f6204441824546b660600/lib/cache_stream.js#L21-L23) seems override `destory` method. Is this correct?

> Tests show that `cacheStream.destroys` in Node.js 14 will be triggered twice, which causes all content to be empty.
> Renaming these two `destroy` methods can solve the problem.

What if the following lines are removed so that `cacheStream.destroy()` is equivalent to `Transform.destroy()`?

https://github.com/hexojs/hexo-util/blob/31f74a5296d74339b76909a05ee38d8bb38560bf/lib/cache_stream.js#L21-L23

**Comment by curbengh** [comment]
> OK. I found this document [transform.destroy([error])](https://nodejs.org/api/stream.html#stream_transform_destroy_error)
> 
> > Implementors should not override this method, but instead implement readable._destroy(). The default implementation of _destroy() for Transform also emit 'close' unless emitClose is set in false.
> 
> But, [hexo-util@ChachStream](https://github.com/hexojs/hexo-util/blob/5b14fd2df882cc3d740f6204441824546b660600/lib/cache_stream.js#L21-L23) seems override `destory` method. Is this correct?

https://github.com/hexojs/hexo-util/pull/195

**Comment by curbengh** [comment]
Since the fixes won't be backported to Hexo 4, I suggest to add `{ "engines" : { "node" : ">=8.10.0 <14" } }` to Hexo 4.2.1. Note that recent npm versions will only show warning but doesn't stop installation in incompatible Node.

If user wants to go ahead with Hexo 4 + Node 14, the fix is to install hexo-util v2 and hexo-fs v3.

**Comment by SukkaW** [comment]
@curbengh 

> I suggest to add `{ "engines" : { "node" : ">=8.10.0 <14" } }` to Hexo 4.2.1

However there are already many features merged into `master` branch. Should we released a `4.3.0.` though?

**Comment by LeoEatle** [comment]
I met the same issue.
Now I change travis.yml to 
```
node_js: 13.2.0
```

**Comment by curbengh** [comment]
> @curbengh
> 
> > I suggest to add `{ "engines" : { "node" : ">=8.10.0 <14" } }` to Hexo 4.2.1
> 
> However there are already many features merged into `master` branch. Should we released a `4.3.0.` though?

I mean a single patch on top of https://github.com/hexojs/hexo/commit/c0bf22284c0c2e85bcd5c94b11823d86a0072035 tagged with [4.2.0](https://github.com/hexojs/hexo/releases/tag/4.2.0).

Other merged changes on top of that commit should be part of 5.0.0, I think some breaking changes are already merged.

**Comment by luisleee** [comment]
I have the same problem.This is the [log](https://api.travis-ci.org/v3/job/682271287/log.txt) of my situation.There are circular dependencie.

**Comment by stevenjoezhang** [comment]
@luisleee circular dependencies are caused by stylus, not Hexo.

**Comment by luisleee** [comment]
OK

**Comment by Rapiz1** [comment]
When will a version compatible with node 14 release?

**Comment by dumindu** [comment]
I [tried even with old node versions like v12](https://github.com/hexojs/hexo/issues/4281#issuecomment-620685855) but still facing issues. Can't we release a hot fix on this?

**Comment by yangxyo** [comment]
> I [tried even with old node versions like v12](https://github.com/hexojs/hexo/issues/4281#issuecomment-620685855) but still facing issues. Can't we release a hot fix for this?
Please run:
```
node --version
```
to show your node version.

**Comment by dumindu** [comment]
@yangxyo Tried both v12.5.0 and v12.16.3 as well. It generates empty files with partial template block codes. You can try https://github.com/hexojs/site

**Comment by luisleee** [comment]
@yangxyo You can reinstall it. It can work on v12.16.3 on my machine.

**Comment by dumindu** [comment]
Oops, my mistake. Looks like current issue is due to https://github.com/hexojs/site/issues/1410 . Fixed. Thanks everyone :)

**Comment by YoshinoriN** [comment]
@LeoEatle @Rapiz1 @dumindu @yangxyo @luisleee 

We published [v4.2.1](https://github.com/hexojs/hexo/releases/tag/4.2.1).
v4.2.1 has included fixed compatibility with Node14.

Thanks :)
